e2e/run_test_cycle.js

```diff
@@ -142,61 +166,69 @@ async function testLoop() {
 
     // Retry on connection/timeout errors
     if (!spec || spec.code) {
-        if (retries >= maxRetryCount) {
-            console.log(chalk.red(`Test ended due to multiple (${retries}) connection/timeout errors with the dashboard server.`));
-            return;
+        if (retry >= maxRetryCount) {
+            return {
+                tryNext: false,
+                count,
+                message: `Test ended due to multiple (${retry}) connection/timeout errors with the dashboard server.`,
+            };
         }
 
-        retries++;
-        console.log(chalk.red(`Retry count: ${retries}`));
-        return testLoop(); // eslint-disable-line consistent-return
-    }
-
-    if (spec.summary) {
-        printSummary(spec.summary);
+        console.log(chalk.red(`Retry count: ${retry}`));
+        return runSpecFragment(count, retry + 1);
     }
 
     if (!spec.execution || !spec.execution.file) {
-        console.log(chalk.magenta(spec.message));
-        return;
+        return {
+            tryNext: false,
+            count,
+            message: spec.message,
+        };
     }
 
     const currentTestCount = spec.summary.reduce((total, item) => {
         return total + parseInt(item.count, 10);
     }, 0);
 
-    console.log(chalk.magenta(`(Testing ${currentTestCount} of ${spec.cycle.specs_registered})  - `, spec.execution.file));
-
-    await runTest(spec.execution, currentTestCount);
+    printSummary(spec.summary);
+    console.log(chalk.magenta(`\n(Testing ${currentTestCount} of ${spec.cycle.specs_registered}) - ${spec.execution.file}`));
+    console.log(chalk.magenta(`At "${process.env.CI_BASE_URL}" server`));
 
-    // Reset retry count on at least one successful run
-    retries = 0;
+    await runCypressTest(spec.execution, count);
 
-    return testLoop(); // eslint-disable-line consistent-return
-}
+    const newCount = count + 1;
 
-function printSummary(summary) {
-    const obj = summary.reduce((acc, item) => {
-        const {server, state, count} = item;
-        if (!server) {
-            return acc;
-        }
+    if (spec.cycle.specs_registered === currentTestCount) {
+        return {
+            tryNext: false,
+            count: newCount,
+            message: `Completed testing of all registered ${currentTestCount} spec/s.`,
+        };
+    }
 
-        if (acc[server]) {
-            acc[server][state] = count;
-        } else {
-            acc[server] = {[state]: count, server};
-        }
+    return {
+        tryNext: true,
+        count: newCount,
+        retry: 0,
+        message: 'Continue testing',
+    };
+}
 
-        return acc;
-    }, {});
+async function runSpec(count = 0, retry = 0) {
```

Comment:
Isn't this still recursion thru async/await operation?

Reply:
Right, still thru async/await, the main difference which I though where the issue is coming from is the [code here](https://github.com/mattermost/mattermost-webapp/pull/7913/files#diff-8686cf1e4a2ca1851e0b683287c8e22cde5d7c441ea88385df53c80893f16722L175):
```javascript
return testLoop(); // eslint-disable-line consistent-return
```
as compared to the [new approach here](https://github.com/mattermost/mattermost-webapp/pull/7913/files#diff-8686cf1e4a2ca1851e0b683287c8e22cde5d7c441ea88385df53c80893f16722R218):
```javascript
const fragment = await runSpecFragment(count, retry);
```

I maybe totally wrong but I tried couple of prod and unstable runs and CircleCI didn't complain/restart the container.